driver-sync/src/main/com/mongodb/client/internal/ClientSessionBinding.java

```diff
@@ -79,23 +79,19 @@ private void closeSessionIfCountIsZero() {
 
     @Override
     public ConnectionSource getReadConnectionSource() {
-        return new SessionBindingConnectionSource(wrapConnectionSource(wrapped.getReadConnectionSource()));
+        if (isActiveShardedTxn()) {
+            return new SessionBindingConnectionSource(wrapped.getConnectionSource(pinServer()));
+        } else {
+            return new SessionBindingConnectionSource(wrapped.getReadConnectionSource());
+        }
     }
 
     public ConnectionSource getWriteConnectionSource() {
-        return new SessionBindingConnectionSource(wrapConnectionSource(wrapped.getWriteConnectionSource()));
-    }
-
-    private ConnectionSource wrapConnectionSource(final ConnectionSource connectionSource) {
-        ConnectionSource retVal = connectionSource;
         if (isActiveShardedTxn()) {
-            setPinnedServerAddress();
-            SingleServerBinding binding = new SingleServerBinding(wrapped.getCluster(), session.getPinnedServerAddress(),
```

Comment:
This is the root of the bug: by creating a new concrete binding here, reaching in to `wrapped` to grab the `Cluster` directly to do so, it subverts any logic contained in `wrapped`.  In the case of client-side encryption, the logic we're skipping is the wrapping of `Connection` in `CryptConnection` by `CryptBinding`, which is why encryption/decryption isn't happening in a sharded transaction.

One way of looking at this bug is that it was caused by a violation of the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle).  Even though this code initially depends on the `ClusterAwareReadWriteBinding` abstraction, it breaks that dependency by depending on a concrete class here.   In that way `CryptBinding` is not substitutable for `ClusterBinding`.

Reply:
The bug is fixed by adding a method to the `ClusterAwareReadWriteBinding` interface that lets this code bind to a specific server.  This allows `ClusterAwareReadWriteBinding` implementations to wrap `Connection` in the way that was intended.